daemon/api_notices.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	noticesCmd = &Command{
+		Path:       "/v2/notices",
+		GET:        getNotices,
+		ReadAccess: openAccess{},
+	}
+
+	noticeCmd = &Command{
+		Path:       "/v2/notices/{id}",
+		GET:        getNotice,
+		ReadAccess: openAccess{},
+	}
+)
+
+func getNotices(c *Command, r *http.Request, user *auth.UserState) Response {
```

Comment:
So I probably went a bit overboard, but I think this function can be made a bit more streamlined. I was thinking of seperating out some of the parsing logic for types so this function becomes a bit simpler. I made a patch for this just to show what I was going for: https://paste.ubuntu.com/p/CGfbhmHHwp/ 

Also there  is no reason to lock the state before parsing the optional value, just in case of an error.

I think maybe one could take it a bit further and also gather all the `parse*` functions into one that reads the query and builds the `NoticeFilter`, but that may be taking it *too far*. Anyway, I hope it's okay I suggested this, it was to avoid putting to many comments.

Reply:
Good point about the state lock, I think fixing that is sufficient reason for a bit of refactoring. Looking ahead a bit, the user-specific notices also have their own helper function for parsing and sanitizing user IDs (akin to types and keys), so I think separating out the types handing into its own function is a good idea as well. I'm not sure we need a separate function to do all the parsing and build the filter, but I'll mention it in the pebble PR I'll open with these changes. The notices code is supposed to be shared between pebble and snapd, with the former as upstream, so I'm going to make sure these suggestions are pebble-approved before committing them here.

Reply:
Merged, and pushed to this PR as well.